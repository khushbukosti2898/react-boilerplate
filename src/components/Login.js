import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  Container,
  Row,
  Col,
  Button,
  Form,
  Card,
  CardBody,
  CardTitle,
} from 'reactstrap';
import { useAuth } from '../hooks/useAuth';
import { setItemInStorage } from '../utils/helper';
import { checkValidation } from './common/form-control/formRules';
import CustomInput from './common/form-control/simple-component/CustomInput';

const initailValue = {
  email: '',
  password: '',
};

const Login = () => {
  const [formData, setFormData] = useState(initailValue);
  const [errors, setErrors] = useState({});
  const onChange = (name, value) => {
    setFormData({ ...formData, [name]: value });
  };

  const validationHandler = (name, error) => {
    setErrors({ ...errors, [name]: error });
  };

  const history = useHistory();
  const auth = useAuth();
  const onLoginClick = (e) => {
    e.preventDefault();
    const validationError = checkValidation(errors, {
      email: formData.email,
      password: formData.password,
    });
    if (Object.keys(validationError).length !== 0) {
      setErrors(validationError);
    } else {
      auth.login();
      setItemInStorage('user', {
        email: formData.email,
      });
      history.push('/');
    }
  };

  return (
    <Container>
      <Row className="justify-content-center align-items-center h-100vh">
        <Col sm="12" md={6}>
          <Card>
            <CardBody>
              <CardTitle tag="h3" className="text-center mb-5">
                Login
              </CardTitle>
              <Form onSubmit={onLoginClick}>
                <CustomInput
                  type="email"
                  name="email"
                  value={formData.email}
                  label="Email"
                  title="email"
                  placeholder="Enter email"
                  isRequired
                  reqType="email"
                  onChange={onChange}
                  validationHandler={validationHandler}
                  error={errors.email}
                  dataTestId="email"
                />
                <CustomInput
                  type="password"
                  name="password"
                  value={formData.password}
                  label="Password"
                  title="password"
                  placeholder="Enter password"
                  isRequired
                  reqType="password"
                  onChange={onChange}
                  validationHandler={validationHandler}
                  error={errors.password}
                  dataTestId="password"
                />
                <Button color="primary mt-2" data-testid="login">
                  Login
                </Button>
              </Form>
            </CardBody>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};

export default Login;
