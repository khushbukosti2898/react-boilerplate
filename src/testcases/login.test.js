import { cleanup, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history'
import { Router } from 'react-router-dom';

import Login from '../components/Login';

const history = createMemoryHistory();

afterEach(cleanup);

describe('Test login component', () => {

  test('Test email is rendered', () => {
    render(<Login />);
    const emailElement = screen.getByTestId('email');
    expect(emailElement).toBeInTheDocument();
    expect(emailElement).toHaveAttribute("type", "email");
  });

  test('Test password is rendered', () => {
    render(<Login />);
    const emailElement = screen.getByTestId('password');
    expect(emailElement).toBeInTheDocument();
  });

  test('Test Login button is rendered', () => {
    render(<Login />);
    const emailElement = screen.getByTestId('login');
    expect(emailElement).toBeInTheDocument();
  });

  test('pass valid email to email input field', () => {
    render(<Login />);
    const emailElement = screen.getByTestId('email');
    userEvent.type(emailElement, "test@mail.com");
    expect(screen.getByTestId('email')).toHaveValue("test@mail.com");
    expect(screen.queryByTestId("input-error")).not.toBeInTheDocument();
  });

  test('pass invalid email to email input field', () => {
    render(<Login />);
    const emailElement = screen.getByTestId('email');
    userEvent.type(emailElement, "test");
    expect(screen.getByTestId("email")).toHaveValue("test");
    expect(screen.queryByTestId("input-error")).toBeInTheDocument();
    expect(screen.queryByTestId("input-error").textContent).toEqual("Please enter valid email.");
  });

  test('pass valid password to password input field', () => {
    render(<Login />);
    const passwordElement = screen.getByTestId('password');
    userEvent.type(passwordElement, "Abc@1234");
    expect(screen.getByTestId("password")).toHaveValue("Abc@1234");
    expect(screen.getByTestId('password').value).toMatch(/Abc@1234/);
    expect(screen.queryByTestId("input-error")).not.toBeInTheDocument();
  });

  test('pass invalid password to password input field', () => {
    render(<Login />);
    const passwordElement = screen.getByTestId('password');
    userEvent.type(passwordElement, "aa");
    expect(screen.getByTestId("password")).toHaveValue("aa");
    expect(screen.getByTestId('password').value).toMatch(/aa/);
    expect(screen.queryByTestId("input-error")).toBeInTheDocument();
    expect(screen.queryByTestId("input-error").textContent).toEqual("Please enter valid password.");
  });

  test('Without fill form click on login', () => {
    render(<Login />);
    expect(screen.getByTestId('email')).toHaveValue("");
    expect(screen.getByTestId('password')).toHaveValue("");
    const loginElement = screen.getByTestId('login');
    userEvent.click(loginElement);
    expect(screen.queryAllByTestId("input-error").length).toBeGreaterThan(0);
  });

  test('valid fill form click on login', () => {
    render(<Router history={history}><Login /></Router>);
    const emailElement = screen.getByTestId('email');
    userEvent.type(emailElement, "test@mail.com");
    const passwordElement = screen.getByTestId('password');
    userEvent.type(passwordElement, "Abc@1234");
    expect(screen.getByTestId('email')).toHaveValue("test@mail.com");
    expect(screen.getByTestId('password')).toHaveValue("Abc@1234");
    const loginElement = screen.getByTestId('login');
    userEvent.click(loginElement);
    expect(screen.queryAllByTestId("input-error").length).toEqual(0);
    expect(history.location.pathname).toBe('/');
  });
});

describe('Snapshot login test', () => {
  test('Snapshot login match', () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
