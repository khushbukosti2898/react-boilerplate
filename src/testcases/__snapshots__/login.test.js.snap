// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot login test Snapshot login match 1`] = `
<div
  className="container"
>
  <div
    className="justify-content-center align-items-center h-100vh row"
  >
    <div
      className="col-sm-12 col-md-6"
    >
      <div
        className="card"
      >
        <div
          className="card-body"
        >
          <h3
            className="text-center mb-5 card-title"
          >
            Login
          </h3>
          <form
            className=""
            onSubmit={[Function]}
          >
            <div
              className="mt-2  form-group"
            >
              <label
                className=""
                htmlFor="email"
              >
                Email
              </label>
              <span
                className="text-danger"
              >
                *
              </span>
              <div
                className="input-group"
              >
                <input
                  checked={false}
                  className="form-control"
                  data-testid="email"
                  disabled={false}
                  id="email"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Enter email"
                  style={Object {}}
                  type="email"
                  value=""
                />
              </div>
              
            </div>
            <div
              className="mt-2  form-group"
            >
              <label
                className=""
                htmlFor="password"
              >
                Password
              </label>
              <span
                className="text-danger"
              >
                *
              </span>
              <div
                className="input-group"
              >
                <input
                  checked={false}
                  className="form-control"
                  data-testid="password"
                  disabled={false}
                  id="password"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Enter password"
                  style={Object {}}
                  type="password"
                  value=""
                />
              </div>
              
            </div>
            <button
              aria-label={null}
              className="btn btn-primary mt-2"
              data-testid="login"
              onClick={[Function]}
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
